infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/utils/LogReplicationConfigManager.java

```diff
@@ -44,95 +68,231 @@ public class LogReplicationConfigManager {
+                    log.error("Cannot open client config tables, retrying.", e);
+                    throw new RetryNeededException();
+                }
+                return null;
+            }).run();
+        } catch (InterruptedException e) {
+            log.error("Failed to open client config tables", e);
+            throw new UnrecoverableCorfuInterruptedError(e);
+        }
+    }
+
+    /**
+     * Generate LogReplicationConfig for all given sessions based on there replication model.
+     *
+     * @param sessions set of sessions for which to generate config
+     */
+    public void generateConfig(Set<LogReplicationSession> sessions) {
+        sessions.forEach(session -> {
+                switch (session.getSubscriber().getModel()) {
+                    case FULL_TABLE:
+                        log.debug("Generating FULL_TABLE config for session {}",
+                                TextFormat.shortDebugString(session));
+                        generateFullTableConfig(session);
+                        break;
+                    case LOGICAL_GROUPS:
+                        log.debug("Generating LOGICAL_GROUP config for session {}",
+                                TextFormat.shortDebugString(session));
+                        generateLogicalGroupConfig(session);
+                        break;
+                    default: break;
+                }
+        });
+    }
+
+    private void generateLogicalGroupConfig(LogReplicationSession session) {
         Map<UUID, List<UUID>> streamToTagsMap = new HashMap<>();
+        Set<String> streamsToReplicate = new HashSet<>();
+        // Check if the local cluster is the Sink for this session. Sink side will honor whatever Source side send
+        // instead of relying on groupSinksMap to filter the streams to replicate.
+        boolean isSink = session.getSinkClusterId().equals(localClusterId);
+        Map<String, Set<String>> logicalGroupToStreams = new HashMap<>();
+        if (!isSink) {
+            groupSinksMap.entrySet().stream()
+                    .filter(entry -> entry.getValue().contains(session.getSinkClusterId()))
+                    .map(Map.Entry::getKey)
+                    .forEachOrdered(group -> {
+                        logicalGroupToStreams.put(group, new HashSet<>());
+                    });
+            log.debug("Targeted logical groups={}", logicalGroupToStreams.keySet());
+        }
 
         registryTableEntries.forEach(entry -> {
-            String tableName = getFullyQualifiedTableName(entry.getKey());
+            String tableName = TableRegistry.getFullyQualifiedTableName(entry.getKey());
+            UUID streamId = CorfuRuntime.getStreamID(tableName);
 
-            if (entry.getValue().getMetadata().getTableOptions().getIsFederated()) {
+            // Add merge only streams first
+            if (MERGE_ONLY_STREAMS.contains(streamId)) {
```

Comment:
this will get executed in every iteration.  Can be moved outside the loop.

Reply:
Not sure I understand, the iteration here is the one for registryTableEntries?

Reply:
merge-only streams contain only the protobuf descriptor and registry table.  They are constants.  Why should we check for them in each iteration.  Please move it outside the loop.

Reply:
I see, I thought you mean handle the MERGE_ONLY streams separately and out of this iteration? I tend to leave it here as otherwise we have to manually add the names and tag to these 2 maps (here the way to get table name is unified). Also, for FULL_TABLE case, it will somehow still need a check there since these 2 table's is_federated flag is false